File: leet-code/49. Group Anagrams/GroupAnagrams49_2.java
package com.example.algorithm;

import java.util.*;

public class GroupAnagrams49_2 {

    public List<List<String>> groupAnagrams(String[] strs) {
        Map<String, List<String>> wordMap = new HashMap<>();
        for (String str : strs) {
            char[] strCharArray = str.toCharArray();
            Arrays.sort(strCharArray);
            String sortedStr = Arrays.toString(strCharArray);
            List<String> words = wordMap.getOrDefault(sortedStr, new ArrayList<>());
            words.add(str);
            wordMap.put(sortedStr, words);
        }
        return new ArrayList<>(wordMap.values());
    }
}

File: leet-code/49. Group Anagrams/GroupAnagrams49_2.kt
class GroupAnagrams49_2 {
    fun groupAnagrams(strs: Array<String>): List<List<String>> {
        val maps: MutableMap<String, MutableList<String>> = mutableMapOf()
        for (str in strs) {
            val sortedStr = str.toCharArray().sorted().joinToString("")
            maps.getOrPut(sortedStr) { mutableListOf() }
            maps[sortedStr]!!.add(str)
        }
        return ArrayList(maps.values)
    }
}

// 참고
// https://www.baeldung.com/kotlin/arraylist-vs-mutablelistof
